### ISSUE TITLE
Port conflict detect function in Windows 10 has a problem

### ISSUE THREAD

**Issue by tianzry** [drakkan/sftpgo#350]
Hi,

Usually, I use sftpgo with portable mode, but I found that sftpgo could not detect port conflict in Windows 10. 

For example, if I have run another FTP server software and use the tcp **port 21**, and then I run the sftpgo with portable mode, setting ftp server **listening port to 21 again**,  however, the sftpgo report no error. This would cause a problem, sftpgo ftp server would fail to respond the request. 

I don't know whether this is Windows 10's feature but it really confuses me sometime, especially when the tcp port I want to use is occupied by another server.

Thank you for your great sftpgo.

**sftpgo version:**
_SFTPGo 2.0.1-cd4a68c-2021-02-06T14:30:52Z +metrics +awskms +gcpkms +vaultkms +azblob +gcs +s3 +bolt +mysql +pgsql +sqlite +portable_

**the command I run sftpgo:**
```
.\sftpgo.exe portable --username root --password test1234 --sftpd-port 2022  --ftpd-port 21 --permissions * --log-file-path ".\sftpgo.log" --directory "D:\temp"
```

the following port usage result was copied from Windows CMD:
```
D:\netstat -ano | findstr 0.0.0.0:21
  TCP    0.0.0.0:21             0.0.0.0:0              LISTENING       15864
  TCP    0.0.0.0:21             0.0.0.0:0              LISTENING       25200
```

**Comment by drakkan** [comment]
Hi,

I'm unable to replicate the issue, please take a look here:

```
Z:\sftpgo_v2.0.2_windows_portable_x86_64>.\sftpgo.exe portable --username root --password test1234 --sftpd-port 445  --ftpd-port 21 --permissions *
2021-03-19T17:33:19.567 INF Portable mode ready, user: "root", password: "[redacted]", public keys: [], directory: "Z:\\sftpgo_v2.0.2_windows_portable_x86_64", permissions: map[/:[*]], enabled ssh commands: [md5sum sha1sum cd pwd scp] file patterns filters: [] SFTP port: 445 FTP port: 21
2021-03-19T17:33:19.581 ERR could not start SFTP server: listen tcp :445: bind: Di norma è consentito un solo utilizzo di ogni indirizzo di socket (protocollo/indirizzo di rete/porta).
```

I tested both port 445 (used by a system service) and port 2022 used by sftpgo running as service. Perhaps the application listening on port 21 uses some kind of port sharing functionality.

Please try to start two sftpgo portable instances using the same ports, the second one should fail

**Comment by tianzry** [comment]
Hi,

Thank you for your reply. More details would be post here, please check.

Before I run FileZilla FTP server, there is no tcp port 21 using:
![image](https://user-images.githubusercontent.com/19747979/111856967-5e347680-8969-11eb-8c35-b8865322efcd.png)


After I run FileZilla FTP server with the tcp port 21, there is ONE tcp port 21 using:

![image](https://user-images.githubusercontent.com/19747979/111856518-40194700-8966-11eb-90c3-688145df99be.png)

![image](https://user-images.githubusercontent.com/19747979/111856992-8c19bb00-8969-11eb-8158-222e4034b528.png)



And then, I run sftpgo with portable mode, using tcp port 21 as ftpd port. Now, there are TWO tcp port 21 using:
```
D:\Tools\sftpgo_v2.0.1_windows_portable_x86_64>.\sftpgo.exe portable --username root --password test1234 --sftpd-port 2022  --ftpd-port 21 --permissions * --log-file-path ".\sftpgo.log" --directory "D:\temp"
2021-03-20T10:18:15.547 INF Portable mode ready, user: "root", password: "[redacted]", public keys: [], directory: "D:\\temp", permissions: map[/:[*]], enabled ssh commands: [md5sum sha1sum cd pwd scp] file patterns filters: [] SFTP port: 2022 FTP port: 21
```
![image](https://user-images.githubusercontent.com/19747979/111856542-5f17d900-8966-11eb-98de-cdcd1849fad8.png)
![image](https://user-images.githubusercontent.com/19747979/111857013-a784c600-8969-11eb-90b8-d6bd7d78aed0.png)



---
And it's same for sftpd port:
Before I run Rebex Tiny SFTP Server with tcp port 2022, there is no tcp port 22 using:
![image](https://user-images.githubusercontent.com/19747979/111856621-f67d2c00-8966-11eb-9cf2-94b218b3b80b.png)

After I run Rebex Tiny SFTP Server with tcp port 2022, there is ONE tcp port 22 using:
![image](https://user-images.githubusercontent.com/19747979/111856646-2c221500-8967-11eb-8d1a-6e6ee2555344.png)
![image](https://user-images.githubusercontent.com/19747979/111856654-3b08c780-8967-11eb-89b6-a5148dfe6ca4.png)

And then I run sftpgo with portable mode, using tcp port 2022 as sftpd port. Now, there are TWO tcp port 2022 using:
```
D:\Tools\sftpgo_v2.0.1_windows_portable_x86_64>.\sftpgo.exe portable --username root --password test1234 --sftpd-port 2022  --ftpd-port 21 --permissions * --log-file-path ".\sftpgo.log" --directory "D:\temp"
2021-03-20T10:30:50.892 INF Portable mode ready, user: "root", password: "[redacted]", public keys: [], directory: "D:\\temp", permissions: map[/:[*]], enabled ssh commands: [md5sum sha1sum cd pwd scp] file patterns filters: [] SFTP port: 2022 FTP port: 21
```
![image](https://user-images.githubusercontent.com/19747979/111856679-625f9480-8967-11eb-947b-7672ae2134f5.png)

![image](https://user-images.githubusercontent.com/19747979/111856695-802cf980-8967-11eb-83b6-dc4e150188c8.png)


---

I didn't run the sftpgo as service, I just run it using command line. 

>Please try to start two sftpgo portable instances using the same ports, the second one should fail
Yes, you are right, in this way, the second one would fail, but we run the same software, the problem is that when sftpgo run  in parallel with other software while using the same tcp port, it cann't  detect the port conflict. So this really confuse me.

My PC's info: Windows 10 ver:1909 
The FileZilla download url: https://filezilla-project.org/download.php?show_all=1&type=server
The Rebex Tiny SFTP Server download url: https://www.rebex.net/tiny-sftp-server/

Thank you again, Drakkan.

**Comment by drakkan** [comment]
Hi,

I'm busy with other things now, I think it could be the same thing reported [here](https://stackoverflow.com/questions/51071020/golang-net-listen-binds-to-port-thats-already-in-use). Do you have time to test?

A quick fix would be to just run SFTPGo instead of Filezilla and Rebex :) I'm kidding of course.

**Comment by tianzry** [comment]
Hi,

Thank you for your reply. 

I'm sorry that I'm not familiar with GO lang, but I try to have a simple test today, following is the result:

First of all, I copy the following code snippet from Internet and the code snippet only opens a socket and waits for connection. I use it for test, so I call it GoSocketTest program. And I will change the parameter of net.Listen() function in line 9.
```go
package main

import (
    "fmt"
    "net"
)

func main() {
    listen, err := net.Listen("tcp", "0.0.0.0:2022")
    if err != nil {
        fmt.Printf("Connection error: %s\n", err)
    } else {
        fmt.Println("Listening")
    }

    coon,err:=listen.Accept()
        if err !=nil{
            fmt.Println("Accept() err=",err)
        }else {
            fmt.Printf("Accept() suc con=%v\n client's ip=%v\n",coon,coon.RemoteAddr())
        }
}
```
---
**Part1:**
Step1: run Rebex Tiny SFTP Server with port 2022,
Step2: check the usage of tcp port 2022
Step3: run GoSocketTest program replacing line 9: `listen, err :=net.Listen("tcp", "0.0.0.0:2022")` 
Step4: check the usage of tcp port 2022 again

The GoSocketTest run well, and there two tcp port 2022 are using:
![image](https://user-images.githubusercontent.com/19747979/111867720-d2463d00-89b0-11eb-8b3a-d0d82f5f70c0.png)


---
**Part2:**
Step1: run Rebex Tiny SFTP Server with port 2022,
Step2: check the usage of tcp port 2022
Step3: run GoSocketTest program replacing line 9: `listen, err :=net.Listen("tcp4", "0.0.0.0:2022")` 

Actually, the GoSocketTest can detect port conflic while filling **tcp4** in net.Listen, so GoSocketTest failed to listen on port 2022 and return error:
![image](https://user-images.githubusercontent.com/19747979/111867399-9c07be00-89ae-11eb-9989-f9d5b8de5719.png)


---
In conclusion, the parameter of network in function net.Listen(network, address) is the key factor, but I dont know why.

**Comment by drakkan** [comment]
Hi,

this is what I suspected, the other services you are running use IPv4 addresses, while Go sockets by default use IPv6 + IPv4 mapped to IPv6. So we have 2 services running on the same port but on different addresses (a plain IPv4 and a IPv4 mapped to IPv6) and so no there is no conflict.

Setting the network to `tcp4` you force the go socket to use an IPv4 address as the other services and so the binding will now fail.

I'll try to better understand why Go sockets work this way and then decide what the best solution/workaround is. I have other priorities for now, it could take a while, sorry.

Thank you for your testing

**Comment by drakkan** [comment]
Hi,

I tryed to report the issue upstream and the answer was "this is working as intended". 
I added a ugly hack, can you please download the Windows build artifact from [here](https://github.com/drakkan/sftpgo/actions/runs/674042855) and confirm that now really works as expected? Thank you

**Comment by tianzry** [comment]
Hi,

Everything looks well now, Thank you.

```
// sftp port 2022 conflict
D:\Tools\Test\sftpgo-windows-latest-go1.16>.\sftpgo.exe portable --username root --password test1234 --sftpd-port 2022  --ftpd-port 25 --permissions * --log-file-path ".\sftpgo.log" --directory "D:\temp"
2021-03-27T16:04:20.641 INF Portable mode ready, user: "root", password: "[redacted]", public keys: [], directory: "D:\\temp", permissions: map[/:[*]], enabled ssh commands: [md5sum sha1sum cd pwd scp] file patterns filters: [] SFTP port: 2022 FTP port: 25
2021-03-27T16:04:20.650 ERR unable to bind tcp4 address: listen tcp4 :2022: bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.

// ftp port 21 conflict
D:\Tools\Test\sftpgo-windows-latest-go1.16>.\sftpgo.exe portable --username root --password test1234 --sftpd-port 2022  --ftpd-port 21 --permissions * --log-file-path ".\sftpgo.log" --directory "D:\temp"
2021-03-27T16:08:49.352 INF Portable mode ready, user: "root", password: "[redacted]", public keys: [], directory: "D:\\temp", permissions: map[/:[*]], enabled ssh commands: [md5sum sha1sum cd pwd scp] file patterns filters: [] SFTP port: 2022 FTP port: 21
2021-03-27T16:08:49.354 ERR unable to bind tcp4 address: listen tcp4 :21: bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.


// sftp and ftp port conflict
D:\Tools\Test\sftpgo-windows-latest-go1.16>.\sftpgo.exe portable --username root --password test1234 --sftpd-port 2022  --ftpd-port 21 --permissions * --log-file-path ".\sftpgo.log" --directory "D:\temp"
2021-03-27T16:04:48.964 INF Portable mode ready, user: "root", password: "[redacted]", public keys: [], directory: "D:\\temp", permissions: map[/:[*]], enabled ssh commands: [md5sum sha1sum cd pwd scp] file patterns filters: [] SFTP port: 2022 FTP port: 21
2021-03-27T16:04:48.966 ERR unable to bind tcp4 address: listen tcp4 :21: bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.
```

**Comment by drakkan** [comment]
Thanks for the confirmation, the only unfortunate thing is that upstream doesn't seem interested in fixing this, so I have to use a hack myself.

I'm not going to backport this change to the 2.0.x branch, it will be included in 2.1.x if it won't cause any other unexpected side effects
